### PR TITLE
Multi Part Download: Use Chunked Stream

### PR DESCRIPTION
**Pull request by GarrettBeatty**
## Description

This PR refactors the buffered multipart download data handling by replacing the single large ArrayPool buffer approach with a chunked buffer stream design. The key changes include:

- **New `ChunkedBufferStream` class**: A writable/readable stream that stores data across multiple small (64KB) ArrayPool buffers, avoiding Large Object Heap (LOH) allocations
- **New `ChunkedPartDataSource` class**: An `IPartDataSource` implementation that wraps `ChunkedBufferStream` for seamless integration with the `PartBufferManager`
- **Removed `BufferedDataSource` and `StreamPartBuffer` classes**: These are replaced by the new chunked approach
- **Updated `BufferedPartDataHandler`**: Modified to use the new `ChunkedBufferStream` instead of `StreamPartBuffer`
- **Updated `PartBufferManager` and `IPartBufferManager`**: Simplified interface and implementation to work with the new data source pattern

## Motivation and Context

The previous buffering implementation used single large ArrayPool buffers which had two issues:
1. **LOH allocations**: Buffers larger than 85KB are allocated on the Large Object Heap, causing memory fragmentation and GC pressure
2. **2GB size limitation**: Single arrays in .NET cannot exceed ~2GB (int.MaxValue bytes), limiting the maximum part size

The new chunked approach:
- Keeps each buffer chunk at 64KB to stay safely below the 85KB LOH threshold
- Supports streams larger than 2GB by distributing data across multiple chunks

https://github.com/aws/aws-sdk-net/issues/3806

## Testing

- Added comprehensive unit tests for `ChunkedBufferStream` (878 lines) covering:
  - Write and read operations
  - Mode switching (write-to-read)
  - Disposal and ArrayPool buffer returns
  - Edge cases and error conditions
- Added comprehensive unit tests for `ChunkedPartDataSource` (602 lines) covering:
  - Construction and validation
  - Read operations
  - Disposal behavior
  - Integration scenarios
- Updated existing `BufferedPartDataHandlerTests` and `PartBufferManagerTests` to work with the new implementation
- Removed obsolete tests for `BufferedDataSource` and `StreamPartBuffer`

### Performance Testing

I did notice some slowdown from the older implementation

**New**
```
Total bytes per run: 5,368,709,120

Run:1 Secs:4.541582 Gb/s:9.456985
Run:2 Secs:8.800254 Gb/s:4.880504
Run:3 Secs:2.014526 Gb/s:21.319988
Run:4 Secs:3.785906 Gb/s:11.344622
Run:5 Secs:11.958203 Gb/s:3.591649
Run:6 Secs:4.626160 Gb/s:9.284086
Run:7 Secs:2.009955 Gb/s:21.368480
Run:8 Secs:1.873424 Gb/s:22.925757
Run:9 Secs:2.329907 Gb/s:18.434070
Run:10 Secs:1.595272 Gb/s:26.923105
```

vs

**Old**
```
Total bytes per run: 5,368,709,120

Run:1 Secs:1.636832 Gb/s:26.239515
Run:2 Secs:1.026863 Gb/s:41.826098
Run:3 Secs:2.580466 Gb/s:16.644151
Run:4 Secs:0.925399 Gb/s:46.412059
Run:5 Secs:0.841140 Gb/s:51.061266
Run:6 Secs:0.816797 Gb/s:52.583070
Run:7 Secs:0.773071 Gb/s:55.557209
Run:8 Secs:0.821784 Gb/s:52.263931
Run:9 Secs:0.816619 Gb/s:52.594513
Run:10 Secs:0.809192 Gb/s:53.077240
```

### 50TB Testing

I also tested with the 50TB file and didnt see any issues

## Screenshots (if appropriate)

Ran using visual studio performance analyzer

before change

image

after change

no LOH allocations

image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Services/S3/Custom/Transfer/Internal/PartBufferManager.cs [hunk 1]
```diff
@@ -176,17 +176,8 @@ internal class PartBufferManager : IPartBufferManager
 
         #endregion
 
-        #region Logger
+        private readonly Logger _logger = Logger.GetLogger(typeof(PartBufferManager));
```
Comment (GarrettBeatty) [inline sdk/src/Services/S3/Custom/Transfer/Internal/PartBufferManager.cs:179]: fixing this to be consistent with the other classes

File: sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs [hunk 2]
```diff
@@ -257,77 +257,61 @@ public void Dispose()
...
-                    }
+                if (chunkedStream.Length != expectedBytes)
+                {
+                    _logger.Error(null, "BufferedPartDataHandler: [Part {0}] Size mismatch - Expected {1} bytes, read {2} bytes",
+                        partNumber, expectedBytes, chunkedStream.Length);
                 }
                 
-                return downloadedPart;
+                // Switch to read mode and wrap in IPartDataSource
+                chunkedStream.SwitchToReadMode();
```
Comment (philasmar) [inline sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs:308]: is there a way to do this automatically? without needing to call it from here

File: sdk/src/Services/S3/Custom/Transfer/Internal/ChunkedBufferStream.cs [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        /// Creates a new ChunkedBufferStream with pre-allocated capacity for the expected size.
+        /// This avoids List resizing during writes, improving performance for known sizes.
+        /// </summary>
+        /// <param name="estimatedSize">The estimated total size in bytes. Used to pre-allocate the chunk list capacity.</param>
+        public ChunkedBufferStream(long estimatedSize)
+        {
+            if (estimatedSize > 0)
+            {
+                // Calculate number of chunks needed and cap at int.MaxValue for List capacity
+                long estimatedChunks = (estimatedSize + CHUNK_SIZE - 1) / CHUNK_SIZE;
```
Comment (philasmar) [inline sdk/src/Services/S3/Custom/Transfer/Internal/ChunkedBufferStream.cs]: This should be estimatedChunks/CHUNK_SIZE, then take the ceiling of the answer

File: sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs (hunk 2 above)
Comment (GarrettBeatty) [reply sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs:308]: updated it to be automatic in the read function. i still kept this code here though to be consistent

File: sdk/src/Services/S3/Custom/Transfer/Internal/ChunkedBufferStream.cs (hunk 3 above)
Comment (GarrettBeatty) [reply sdk/src/Services/S3/Custom/Transfer/Internal/ChunkedBufferStream.cs]: updated to use original formula but added docs as discussed

**Comment by GarrettBeatty** [comment]
> The code looks fine but I am concerned about the performance drop. Could it be that the chunk size is too small and it's creating a lot of overhead? Could you do some testing to increase the chunk size to int.MaxValue (disregarding the LOH optimization) and see if we regain the old performance? If that's the case, maybe we need to find the sweet spot for the chunk size instead of trying to completely avoid the LOH?

talked on devex channel internally on why its okay to use 64KB default

**Review by peterrsongg — APPROVED** [review]
